### PR TITLE
fix(test-fixtures): relocate Scratch dirs instead of panicking under a carve-out

### LINKED ISSUE #403: feedback(cadence-hooks): the test suite refuses a /tmp worktree, conflicting with the documented worktree pattern
## Problem

The house worktree pattern puts feature worktrees in `/tmp`, but this repo's own test suite **hard-refuses a temp fixture root**. Following the convention makes `cargo test --workspace` fail 66 tests in crates the change never touched, which reads exactly like the change broke the suite.

## Environment

- `cadence-hooks 0.67.0`
- macOS 26.5.1

## Reproduction

```bash
git -C <cadence-hooks> worktree add /tmp/my-feature -b my-feature origin/main
cd /tmp/my-feature
cargo test --workspace --no-fail-fast
```

Result — 66 failures:

- `crates/core`: 5 failed (`worktree::tests::*`)
- `crates/guardrails`: 61 failed (`enforce_worktree::tests::*`)

Every one panics with the same assertion:

```
fixture root sits under a carve-out (.claude/ or temp) — run the suite from a carve-out-free checkout:
/private/tmp/my-feature/crates/core/../../target/core-worktree-scratch/would-block-...
```

Relocating the worktree to `~/Projects/` and re-running gives a fully green workspace.

## Why this is worse than a plain env requirement

The failures name `crates/core` and `crates/guardrails` — crates a typical feature diff never touches. The first read is "my change broke 66 tests in unrelated crates," and the assertion text is the only thing that says otherwise. The convention that put the worktree there is documented elsewhere, so nothing at the failure site connects the two.

## Secondary: `git worktree move` silently invalidates compiled test binaries

Relocating out of `/tmp` with `git worktree move` produced a second, unrelated-looking failure wave — every integration test binary failing with:

```
failed to execute binary: Os { code: 2, kind: NotFound, message: "No such file or directory" }
```

`CARGO_BIN_EXE_*` is baked in at **compile time**, so the moved test binaries still pointed at the pre-move absolute path. Cargo does not detect this and will not rebuild, because the sources are unchanged. `cargo clean` is required after a worktree move. Worth a line wherever worktree relocation is described.

## Expected vs actual

- **Expected:** a worktree created by the documented pattern can run the repo's own test suite.
- **Actual:** the documented location is the one location the suite refuses, and the failure presents as unrelated-crate breakage.

## Lane

Friction — generalizable, not specific to this session's shape. Any feature branch in this repo that follows the worktree convention hits it on the first `cargo test`.

## Suggested seam (not a design)

Two candidate resolutions, both yours to pick between: relax the fixture-root assertion so a temp root is usable, or document the carve-out conflict where the worktree pattern is taught and at the failure site. Reporting the seam only.

### PR DESCRIPTION
**Pull request by cameronsjo**
Closes #403

Test fixtures build their paths from `CARGO_MANIFEST_DIR`, which lives *inside* the checkout. When the checkout itself sits under an enforce-worktree carve-out — a `/tmp` worktree, following this repo's own documented convention — every fixture path inherits that prefix and trips the guard's own `is_claude_managed_dir` / `is_temp_root` exemption. Verdicts that should BLOCK come back ALLOW, and the block-expecting tests fail on a genuinely correct branch.

## Measured, with a positive control

**92 failures**, `cargo test --workspace --no-fail-fast` exit 101, at `76068a6` (`origin/main`):

| crate | failures |
|---|---|
| `tests/deadline_failopen.rs` | 2 |
| `cadence-hooks-core` | 11 (`git_fixtures::tests::*`, `worktree::tests::*`) |
| `cadence-hooks-guardrails` | 76 (74 `enforce_worktree::tests::*` — nearly the whole module — plus 2 `warn_branch_base::tests::*`) |
| `cadence-hooks-session` | 3 (`start::tests::posture_line_fires_in_primary_checkout`, `run_start_nudges_on_posture_alone_with_zero_peers`, `run_start_joins_posture_and_peer_disclosure`) |

**Positive control:** the identical suite, same SHA, same build steps, in a worktree *not* under `/tmp` or `.claude/` — **3758 run, 3758 passed, exit 0.** Same total test count in both runs, so it is demonstrably the same suite with only the carve-out condition differing. Without that control the 92 would be a number, not evidence.

Three measurement traps were hit and are worth recording, because each would have produced a confidently wrong figure:

- **`cargo test` fail-fasts by default.** A plain `cargo test --workspace` stops after the first failing test binary and surfaced **2** failures. `--no-fail-fast` is what produces the real total.
- **`CADENCE_ALLOW_MAIN=true` was ambiently set** in the environment. It silently exempts every enforce-worktree check, so a broken suite reads green. Both runs used `env -u CADENCE_ALLOW_MAIN -u CADENCE_NO_ENFORCE_WORKTREE`.
- **`cargo test` does not build the `cadence-hooks` binary.** Pipe-probing tests exit 127, which misreads as a false ALLOW. `cargo build --bin cadence-hooks` ran first in both checkouts.

An earlier in-lane measurement reported 90 against a repro worktree that was two commits behind with four dirty files — a partially-fixed tree. The 92 above was taken after confirming a clean checkout at a verified SHA. Reported rather than silently reconciled, because the discrepancy is exactly the shape of the "never measure against a working tree" trap.

## Relationship to #494 — it did not cover this; it widened it

#494 promoted a `target/`-rooted `Scratch` with a hard carve-out assert into `core::git_fixtures`, on the reasoning that a `target/`-relative root sits outside the carve-outs. **That holds only when the ambient checkout is not itself carved out.** Under a `/tmp` worktree, `target/` inherits the same prefix and the assert panics instead of protecting anything — so #494 is the direct cause of most of the 92 failures, not prior art that solved part of this.

`session::start` additionally carried its own independent hand-rolled `Scratch` with **no** carve-out check at all — silently wrong rather than panicking, a different failure mode #494's promotion never reached. That is #437's root cause, and a comment already in `start.rs` describes the symptom.

## Approach, and why the alternatives lost

**Chosen: generalized relocation.** `resolve_scratch_dir` tries the default `target/`-relative root first (unchanged for the common case), then falls back to `$CADENCE_HOOKS_TEST_SCRATCH_ROOT`, then `$CARGO_HOME` / user-home + `.cargo`, panicking only when nothing escapes.

- **"Make the fixture root non-exempt"** is what #494 already attempted, and it is structurally incapable of working: the root is computed from inside the checkout, so when the checkout is carved out, *no subdirectory of it can escape*. Fixing it requires leaving the checkout entirely, which is what the fallback chain does.
- **Detect-and-skip** trades a confusing failure for a silent coverage gap. Every guard's real block/allow logic would go untested from a `/tmp` checkout — precisely where #226 and #437 show people actually run the suite.
- **Docs-only** leaves the trap in place.

## The trade this makes, stated rather than left implicit

#312's assert caught **two** things: a genuinely carved-out checkout (this bug), and a caller passing a bad `root` (a misuse `worktree.rs` still documents). The resolver cannot distinguish them, so caller error is now **silently repaired** rather than reported.

Accepted deliberately. The correctness property survives — a relocated fixture is still non-exempt, so nothing passes vacuously — but the caller-error signal is genuinely gone, and that is a real cost rather than a technicality.

## Review findings folded

- **`cargo_home()` was a degraded fork of `crate::paths::user_home()`** — missing the Windows `USERPROFILE`/`HOMEDRIVE`+`HOMEPATH` chain (and `CONTRIBUTING.md` documents Windows support), and missing the empty-as-unset filter. With `HOME=""` it yielded the **relative** path `.cargo`, which passes the carve-out check and gets created under the cwd — a silent wrong answer in the one branch written to fail loudly. The same empty-var filter was applied to `$CARGO_HOME` itself, which had the identical defect one layer up.
- **Two `Scratch`-based relocation tests removed.** They passed hardcoded carve-out roots and therefore relocated on *every* run, writing into a real `$CARGO_HOME`. Confirmed empirically: `~/.cargo/cadence-hooks-test-scratch` existed on the dev machine, created from a worktree under no carve-out. **The justification for removing them — that the hermetic tests already pinned every branch — was not true when it was written:** no test passed `cargo_home: Some(carved-out-path)` with `override_root: None`, leaving that branch correct-but-unpinned. It is pinned now by `resolve_panics_when_cargo_home_itself_is_carved_out_and_no_override`, and the chain is covered by **eight** tests rather than six — the six original branches plus that one and a real-symlink test.
- **The relocation note fires once per test binary**, behind a `std::sync::Once`, rather than once per fixture (~120 call sites). `CONTRIBUTING.md` says "once per test binary" rather than implying a single global line, since `cargo test --workspace` spawns separate processes with no shared state.
- **`$TMPDIR` canonicalized once** and reused across all three candidates, rather than a `realpath` chain per candidate — on precisely the relocation path this fix exists to serve.
- `create_dir_all` now names `CADENCE_HOOKS_TEST_SCRATCH_ROOT` on failure, so a read-only or vendored `$CARGO_HOME` reports the escape hatch instead of a bare `ReadOnlyFilesystem`. Parameter types made symmetric (`Option<&Path>`), `SCRATCH_ROOT_OVERRIDE_ENV` and `resolve_scratch_dir` reduced to `pub(crate)`, and a test renamed to match what it actually asserts.

## `CLAUDE.md`

Two entries described traps this change resolves. Both compressed to one line each, and **the still-live pipe-probe trap was split out into its own present-tense entry** — it had been buried inside an entry headed `RESOLVED`, where a reader scanning for active gotchas would stop at the heading and never reach a live false-ALLOW misread.

Net: `CLAUDE.md` is **919 bytes smaller** than before this change (19,306 vs 20,225), not larger.

## Verification

```
cargo test --workspace                                  exit 0, 0 failed
  — from a non-carve-out worktree AND a fresh /tmp worktree
cargo clippy --workspace --all-targets -- -D warnings   exit 0
cargo fmt --all -- --check                              exit 0
git status --porcelain                                  clean
```

**Seeded-regression check:** `should_block` was flipped to ignore `allowed_main` and the suite re-run from the `/tmp` worktree — **16 real failures surfaced**, confirming the relocation does not mask genuine breakage. Reverted after.

Confirmed from the committed blob via `git show HEAD:<path>`, not the working tree.

## Deferred, not dropped

**#504** — three fixture duplications that survived the `git_fixtures` sweep, including two copies of the `scratch_root()`/`scratch(tag)` boilerplate this change itself adds. **#505** — whether the fallback root belongs in `$CARGO_HOME` or `$XDG_CACHE_HOME`, given `cargo clean` never sweeps the fallback.

## Note on this branch's history

The first commit is a `wip:` checkpoint the orchestrator committed on the implementing lane's behalf while the work lived only in a working tree. It carries no reviewed content and is superseded by the commits above it.

## Review trail, stated precisely

CodeRabbit posted a real review at `e7c8eed` (5 actionable comments, all folded — see above). **It did not review the two commits after that.** Its last activity was 20:09Z; the head `6345ea7` was pushed at 21:21Z, so the green "Review completed" check covers an older commit. That is worth naming: a stale CodeRabbit review and a rate-limited one both present as a green check.

The unreviewed delta `e7c8eed..6345ea7` is two test-only commits — a `#[cfg(unix)]` import gate, and the temp-dir anchoring above. Both CI legs pass on the head, **including Windows**, which is the only oracle for the platform half. The orchestrator reviewed the delta directly against the one risk that matters for a platform fix: whether the five tests were *fixed* or *neutered*.

Traced per test, and the answer is fixed:

- The default candidate is genuinely rejected in each — anchoring it under `$TMPDIR` *and* passing that value as `tmpdir` routes through the portable `via_env` branch, so control actually reaches the override / `$CARGO_HOME` / panic path each test names.
- Each can still fail. `resolve_override_wins_when_default_is_carved_out` passes `/Users/dev/.cargo/should-not-be-used` as `cargo_home` specifically so that removing the override branch lands there and fails the assertion — a built-in discriminator between "override won" and "something else won."
- `resolve_falls_back_to_cargo_home_when_default_is_claude_managed` was deliberately **not** re-anchored: `.claude` is a literal directory name, so that test was already platform-independent.
- No filesystem or concurrency risk from the new helpers: `resolve_scratch_dir_with` is pure — it computes paths and never creates them — so the six distinct tags never touch disk. The `temp_dir()` read sits in a test helper, not in the function under test, so the injectable-parameter boundary the module exists to protect is unchanged.

## One correction to this branch's history

The commit message on `6345ea7` says the implementing lane "went silent before committing." **That is false.** The lane was mid-flight; the orchestrator's status snapshot landed in the gap, and its own work was committed on its behalf with content identical to what it was about to push. Nothing was lost, but the claim is wrong and is corrected here rather than by rewriting pushed history.

It also exposed a real gap in the give-up rule that shipped earlier the same day: the rule says the second re-engage is the last action and "continued silence past it" is the give-up point, but never states how long that window is — so the orchestrator re-engaged and gave up seven minutes later. Filed as a sixth finding on cameronsjo/cadence#685.

Session-Name: tidal-viola
Session-Id: ecc24a74-e772-4644-b6ed-342b00644f0b
Model: claude-opus-5
Harness: claude-code 2.1.215
Machine: cf6e768835c7

### CODE REVIEW

File: crates/core/src/git_fixtures.rs [hunk 1]
```diff
@@ -73,6 +89,136 @@ impl Scratch {
...
+    let candidate = |base: &Path| base.join(&suffix);
+    let escapes_carveout = |dir: &Path| {
+        !is_claude_managed_dir(dir)
+            && !path_under_temp_root_with_canonical(dir, tmpdir, tmpdir_canonical.as_deref())
+    };
+
+    let default_dir = candidate(root);
+    if escapes_carveout(&default_dir) {
+        return default_dir;
+    }
```
[the remark this answers is not in the document]
Comment (cameronsjo) [reply crates/core/src/git_fixtures.rs:164]: Verified this reproduces before fixing, as requested. Built a real symlink at a non-carve-out path pointing into `/tmp`, and confirmed `git rev-parse --show-toplevel` run from inside it resolves to the real `/private/tmp/...` path, not the symlinked literal path — the same resolution `enforce-worktree`'s own checks go through. So a lexical-only check on the literal path really could diverge from what the guard being tested sees.

Fixed in `e7c8eed`: added `canonicalize_nearest_existing` (walks up to the nearest existing ancestor — the `{tag}-{pid}` leaf doesn't exist yet when this check runs — canonicalizes, then re-appends the removed tail) and route the candidate directory through it before both `is_claude_managed_dir` and `path_under_temp_root_with_canonical`. Added `resolve_scratch_dir_resolves_symlinks_before_checking_the_carveout`, which builds a real symlink and asserts the fallback fires.

File: crates/core/src/git_fixtures.rs [hunk 2]
```diff
@@ -73,6 +89,136 @@ impl Scratch {
...
+        return dir;
+    }
+
+    if let Some(cargo_home) = cargo_home {
+        let dir = candidate(&cargo_home.join("cadence-hooks-test-scratch"));
+        if escapes_carveout(&dir) {
+            note_relocation_once(&format!("to $CARGO_HOME: {}", dir.display()));
+            return dir;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (cameronsjo) [reply crates/core/src/git_fixtures.rs:204]: Fixed in `e7c8eed`: the `$CARGO_HOME`/override fallback bases now append `root`'s own final path component (`git-fixtures-scratch`, `enforce-worktree-scratch`, `session-posture-scratch`, ...) as a namespacing subdirectory before joining the `{tag}-{pid}` leaf, so distinct callers no longer collapse into one shared directory. Confirmed the existing `starts_with` assertions in the affected tests still hold (they check a prefix, not an exact path).

File: crates/core/src/git_fixtures.rs [hunk 3]
```diff
@@ -73,6 +89,136 @@ impl Scratch {
...
+/// `PathBuf::from("")` — and `.join(".cargo")` on it resolves the *relative*
+/// path `.cargo`, which names neither `.claude/` nor a temp root, so
+/// `escapes_carveout` would bless it and this fixture would be created
+/// relative to the test binary's cwd instead of failing loudly.
+fn cargo_home() -> Option<PathBuf> {
+    std::env::var_os("CARGO_HOME")
+        .map(PathBuf::from)
+        .filter(|p| !p.as_os_str().is_empty())
+        .or_else(|| crate::paths::user_home().map(|h| h.join(".cargo")))
+}
```
[the remark this answers is not in the document]
Comment (cameronsjo) [reply crates/core/src/git_fixtures.rs:297]: Fixed in `e7c8eed`. Made `crate::paths::non_empty_var` `pub(crate)` and reused it in `cargo_home()` instead of hand-rolling the empty-as-unset filter again. Split the composition into a pure `cargo_home_from(cargo_home_var: Option<String>, home: Option<PathBuf>) -> Option<PathBuf>`, so the rule (env var wins if non-empty, else `home.join(".cargo")`, else `None`) is pinned by 3 tests with no process-env mutation:  `cargo_home_from_prefers_the_env_var`, `cargo_home_from_falls_back_to_home_when_env_var_is_none`, `cargo_home_from_is_none_when_neither_is_available`.

Also closed a gap found while doing this: no existing test passed `cargo_home` as itself carved-out (only `None`), so that branch of the final panic was correct but unpinned — added `resolve_panics_when_cargo_home_itself_is_carved_out_and_no_override`.

File: crates/core/src/git_fixtures.rs [hunk 4]
```diff
@@ -171,19 +352,101 @@ mod tests {
...
+    fn resolve_falls_back_to_cargo_home_when_default_is_claude_managed() {
+        let dir = resolve_scratch_dir_with(
+            Path::new("/Users/dev/checkout/.claude/scratch"),
+            "tag",
+            None,
+            None,
+            Some(Path::new("/Users/dev/.cargo")),
+        );
+        assert!(!is_claude_managed_dir(&dir));
+    }
```
[the remark this answers is not in the document]
Comment (cameronsjo) [reply crates/core/src/git_fixtures.rs:550]: Fixed in `e7c8eed`. Added a `starts_with("/Users/dev/.cargo/cadence-hooks-test-scratch")` assertion (mirroring the sibling temp-root test) alongside the existing `!is_claude_managed_dir` check, so the test now pins that the $CARGO_HOME fallback branch actually fired, not just that the result happens not to be under `.claude/`.

File: crates/core/src/worktree.rs [hunk 5]
```diff
@@ -139,8 +139,35 @@ pub(crate) fn git_dir_is_common_dir(git_dir: &Path, repo_root: &Path) -> bool {
...
+    path_under_temp_root_with_canonical(path, tmpdir, canonical.as_deref())
+}
+
+/// `$TMPDIR`, canonicalized once — factored out of [`path_under_temp_root`]
+/// so a caller checking multiple candidate paths against the same `$TMPDIR`
+/// (e.g. [`crate::git_fixtures::resolve_scratch_dir`]'s up-to-three-candidate
+/// fallback chain, cadence-hooks#403 code review) can canonicalize once and
+/// reuse it, instead of re-running a real `realpath` syscall chain per
+/// candidate.
+pub fn canonicalize_tmpdir(tmpdir: Option<&str>) -> Option<PathBuf> {
```
[the remark this answers is not in the document]
Comment (cameronsjo) [reply crates/core/src/worktree.rs]: Fixed in `e7c8eed`: `canonicalize_tmpdir` and `path_under_temp_root_with_canonical` narrowed from `pub` to `pub(crate)` — confirmed no caller outside `crates/core` (only `crates/core/src/worktree.rs` and `crates/core/src/git_fixtures.rs` reference them).
